components/DiffView.tsx

```diff
@@ -0,0 +1,104 @@
+import React, { useEffect, memo } from 'react'
+import gitDiffParser, { File } from 'gitdiff-parser'
+import Prism from 'prismjs'
+import ReactDiffViewer from 'c0d3-diff'
+import _ from 'lodash'
+import CommentBox from './CommentBox'
+import { Comment } from '../graphql'
+
+const prismLanguages = ['js', 'javascript', 'html', 'css', 'json', 'jsx']
+
+const DiffView: React.FC<{
+  diff?: string
+  id: number
+  name: string
+  username: string
+  comments: Comment[]
+  lessonId?: number
+  status?: string
+}> = ({ diff = '', id, name, username, comments, lessonId, status }) => {
+  const files = gitDiffParser.parse(diff)
+  type fileComments = Record<string, { lines?: number[]; comments?: Comment[] }>
+  //every file gets unique index in format of submissionId:fileName
+  const [commentsState, setCommentsState] = React.useState<fileComments>({})
+
+  useEffect(() => {
+    const commentMap: fileComments = {}
```

Comment:
This should be refactored to use `reduce` instead of `forEach`, because you are using `forEach` to do what `reduce` already does. This change will make the code more understandable,

```
const commentMap = comments && comments.reduce(...)
```

Reply:
I changed `forEach` to reduce.